### ISSUE TITLE
Forced locale switch with no possibility to change it to English

### ISSUE THREAD

**Issue by rumatoest** [rust-lang/rust-www#623]
Hello.

Ok if you think (or guess) that many know Russian it does not meant that you should force your web site to show me content on that language.

The problem: 
 - I go to https://www.rust-lang.org
 - it redirects me to https://www.rust-lang.org/ru-RU
 - Why I should waste my time to scroll down and find language switch there?
 - Why After switch to normal (English) language all my next clicks on links will redirect me at /ru-RU/** ?

I'm really appreciate your help. But don't. Please. I just want to read English site as well as many other Russian speaking developers.

**Comment by Centaur** [comment]
Yes. This behavior is very annoying, especially the 4th point, please fix it.

**Comment by Centaur** [comment]
Ok. Now I see there already have been several issues on this problem here but no maintainer has ever shown up. Shame.

**Comment by lifthrasiir** [comment]
I believe this is ultimately a symptom from using a completely static website. (I've done it for my personal website multiple times, and it is very frustrating to get it right! [1]) So we have at least three problems from this root problem:

* There is no language-neutral canonical URL readily accessible from the location bar. One needs to manually "fix" the URL to get one---no one does that.
* The user language preference for the website (as opposed to the global settings) is not saved.
* The localized website has to carefully link against other pages to avoid linking to the wrong language, whether it's a correct choice or not.

It is possible to somehow mitigate the issue with JavaScript, but in my experience it has never been perfect enough. I don't think S3 (that's where the current website is hosted AFAIK) is flexible enough to support this use case---probably a thin web server backend is better, though this means that there has to be a minimal load balancer. (The web server itself can be a nginx plus a moderately complex configuration. The website would remain static and the language preference would be stored in a first-party cookie.) I think I'm not qualified to say that it's a way to go, though.

[1] My "current" approach has been to mimic the outlined procedure with the [History API](https://developers.whatwg.org/history.html), which is now [widely supported](http://caniuse.com/#feat=history), except for the dreaded IE 9 or before. I had tons of fallbacks to get it accessible for unsupported devices, and had to account for the presence of query strings, fragment identifiers, in-page navigations and so on.

**Comment by steveklabnik** [comment]
> Now I see there already have been several issues on this problem here but no maintainer has ever shown up. Shame.

There's been a couple of issues, yes, but it's also not 100% clear what to do to fix this yet, exactly. @lifthrasiir does a good job of going into some of the details.

**Comment by yeputons** [comment]
@rumatoest I believe that the commit has landed. Would you mind confirming that the problem is resolved?

**Comment by rumatoest** [comment]
Now it remembers my language choice, but it does not work when I go to main page https://www.rust-lang.org - still reset it to default.

BTW language switch at site footer is not so good idea.

**Comment by yeputons** [comment]
@rumatoest did I understand you correctly that "but it does not work when I go to main page" refers to the fact that default website's language is chosen based on your browser/OS' preferences and its not English for all cases? If it's the only problem left, I think it'd be better to rename the issue to something like "Consider switching off default locale based on browser/os settings". If you also do not like language switch at site footer, I'd create a separate issue with that.

**Comment by rumatoest** [comment]
> and its not English for all cases?

When I first go to main page - it switch language to Russian.
I switch language to English
Now I can view  other pages on English
But when I open site main page it switch me to Russian again.

Why you just do not use Cookie to save locale settings?

> If you also do not like language switch at site footer, I'd create a separate issue with that.
It will be nice.

**Comment by uppi** [comment]
Moreover, when you switch to English, you can browse the site in English until you meet any page that is not translated to Russian (e. g. https://www.rust-lang.org/conduct.html). 
It resets the language choice and redirects you to https://www.rust-lang.org/ru-RU/conduct.html, which is a page in Russian, telling you that the requested page was not translated.
